Convert "Reasoning with inductive types" chapter to prodn. (#14179)
coq/coq: doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

   .. coqtop:: reset none

      Goal  forall P1 P2 : Prop, P1 -> P1 \/ P2.

   .. coqtop:: all

      intros; apply or_introl.  (* equivalent to "left" *)

.. _CaseAnalysisAndInduction:

Case analysis
-------------

The tactics in this section implement case
analysis on inductive or co-inductive objects (see :ref:`variants`).

.. tacn:: destruct {+, @induction_clause } {? @induction_principle }

   .. insertprodn induction_clause induction_principle

   .. prodn::
      induction_clause ::= @induction_arg {? as @or_and_intropattern } {? eqn : @naming_intropattern } {? @occurrences }
      induction_principle ::= using @one_term {? with @bindings } {? @occurrences }

   Performs case analysis by generating a subgoal for each constructor of the
   inductive or co-inductive type in the :n:`@one_term` in :n:`@induction_arg`.
   In each new subgoal, the tactic replaces :n:`@one_term` with the associated
   constructor (applied to its arguments).   Unlike :tacn:`induction`,
   :tacn:`destruct` generates no induction hypothesis.

   The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
   the tactic checks that all subterms matching the pattern in the conclusion

>    and the hypotheses are compatible and performs case analysis using this subterm.

Reviewer: I recently used this feature in a proof and this description is wrong: there is no check that all subterms matching the pattern are compatible. The tactic just selects the first subterm matching the pattern that it finds and performs case analysis using this subterm.

Example:
```coq
Require Import Arith.

Goal forall x y, Nat.compare x y = Nat.compare y x.
intros.
destruct (Nat.compare _ _).
```

      defined.  If there are not enough names, Coq picks fresh names.

   :n:`eqn : @naming_intropattern`
      Generates a new hypothesis in each new subgoal that is an equality between
      the term being case-analyzed and the associated constructor (applied to
      its arguments).  The name of the new item may be specified in the
      :n:`@naming_intropattern`.

   :n:`with @bindings`  (in :n:`@one_term_with_bindings`)
      Provides explicit instances for
      the :term:`dependent premises <dependent premise>` of the type of
      :token:`one_term`.

   :n:`@occurrences`
     Selects specific subterms of the goal and/or hypotheses to apply
     the tactic to.  See :ref:`Occurrence clauses <occurrenceclauses>`.
     If it occurs in the :n:`@induction_principle`, then
     there can only be one :n:`@induction_clause`, which can't have its
     own :n:`@occurrences` clause.

   :n:`@induction_principle`
     Makes the tactic equivalent to
     :tacn:`induction` :n:`{+, @induction_clause } @induction_principle`.

   .. tacn:: edestruct {+, @induction_clause } {? @induction_principle }

      Equivalent to :n:`destruct @term` except that it does not

>       fail if the instance of a dependent premises of the type

Reviewer: ```suggestion
      fail if the instance of a dependent premise of the type
```

      of :token:`term` is not inferrable. Instead, the unresolved instances
      are left as existential variables to be inferred later, in the same way
      they are in :tacn:`eapply`.

.. tacn:: case {+, @induction_clause } {? @induction_principle }

   .. todo PR: clear as mud? (below) the syntax accepts more parameters than elim
      not sure what to say here

   An older, more basic tactic to perform case analysis without
   recursion.  We recommend using :tacn:`destruct` instead where possible.
   `case` only modifies the goal; it does not modify the :term:`local context`.

   .. tacn:: ecase {+, @induction_clause } {? @induction_principle }

      If the type of :n:`@one_term` (in :n:`@induction_arg`) has
      :term:`dependent premises <dependent premise>`, or dependent premises
      whose values are not inferrable from the :n:`with @bindings` clause,

>       :n:`ecase` turns them into existential variables to be resolved later on.

Reviewer: This explanation is much better than the one for `edestruct`. Can you copy it over there?

   of the underlying technique :ref:`here <dependent-induction-examples>`.

Induction
---------

.. tacn:: induction {+, @induction_clause } {? @induction_principle }

   Applies an :term:`induction principle` to generate a subgoal for
   each constructor of an inductive type.

   If the argument is :term:`dependent <dependent product>` in the conclusion or some
   hypotheses of the goal, the argument is replaced by the appropriate
   constructor in each of the resulting subgoals and induction
   hypotheses are added to the local context using names whose prefix
   is **IH**.  The tactic is similar to :tacn:`destruct`, except that
   `destruct` doesn't generate induction hypotheses.

   There are several cases:

   + If :n:`@one_term` (in :n:`@induction_arg`)
     is an identifier :n:`@ident` denoting a quantified variable of the
     conclusion of the goal, then :n:`induction @ident` behaves like :tacn:`intros` :n:`until
     @ident; induction @ident`. If :n:`@ident` is no longer dependent in the
     goal after application of :n:`induction`, it is erased. To avoid erasure,
     use parentheses, as in :n:`induction (@ident)`.

>    + If :n:`@one_term` is a :n:`@natural`, then :n:`induction @natural` behaves like

Reviewer: ```suggestion
   + If :n:`@induction_arg` is a :n:`@natural`, then :n:`induction @natural` behaves like
```

   :tacn:`destruct` generates no induction hypothesis.

   The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
   the tactic selects the first subterm that matches the pattern and performs
   case analysis using that subterm.

   :n:`{+, @induction_clause }`
     Giving multiple :n:`@induction_clause`\s is equivalent to applying :n:`destruct`
     serially on each :n:`@induction_clause`.

   :n:`@natural`
     Equivalent to :tacn:`intros` :n:`until @natural` followed by `destruct` applied to
     the last introduced hypothesis.

     .. note::
        To destruct a number, use the syntax :n:`destruct (@natural)` (not
        very interesting anyway).

   :n:`@ident` (in :n:`@induction_arg`)
     If :token:`ident` is a quantified variable in the conclusion
     of the goal, then :n:`destruct @ident` behaves like :tacn:`intros`
     :n:`until @ident; destruct @ident`.
     If :token:`ident` is a hypothesis of the context and :token:`ident`
     is no longer dependent in the goal after application
     of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as
     in :n:`destruct (@ident)`).

   :n:`as @or_and_intropattern`

>       Provides names for the variables introduced in each new subgoal.  The

Reviewer: It could be clarified that the inner intropatterns are not restricted to naming intropatterns.

Author: Is adding this at the end sufficient?

The inner :n:`intropattern`s can also split introduced hypotheses into
multiple hypotheses or subgoals.

Reviewer: This goes beyond that: they can also be used to do rewriting (`->` and `<-`), we can apply lemmas to the introduced hypotheses `%` or injection (`[= ]`)... Basically, the full power of intropatterns is available for inner intropatterns (but only for inner intropatterns).

Author: Is it really necessary to repeat everything that `intropattern` can do given that text already hyperlinks to the `intropattern` nonterminal?  And if so, what wording do you suggest?

Reviewer: No, it would be a terrible idea to repeat everything that intro patterns can do because it would quickly become outdated. I just suggest to clarify that not just naming patterns are supported but the whole set (as long as they are nested).

Author: Then how about:

> Note the inner :n:`intropattern`s are not restricted to renaming; they can be used to split introduced hypotheses into multiple hypotheses, to do rewriting (`->` and `<-`), to apply lemmas to the introduced hypotheses (`%`) or to do `injection` (`[= ... ]`).

I really don't understand why you want this tactic to have this extra detail that's covered in in the intropattern section.  If you don't like this wording, please provide specific wording.

Reviewer: What you propose is fine.

>I really don't understand why you want this tactic to have this extra detail that's covered in in the intropattern section.

**I don't!** I just want to clarify that intropatterns are not semantically restricted to naming. This is what is suggested by the current sentence "Provides names for the variables introduced in each new subgoal."

The change could have been as simple as:

>Provide names for (or apply further transformations to) the introduced variables and hypotheses in each subgoal.

Author: Got it.  It is now:

Provides names for (or apply further transformations to)
the variables and hypotheses introduced in each new subgoal.

      the term being case-analyzed and the associated constructor (applied to
      its arguments).  The name of the new item may be specified in the
      :n:`@naming_intropattern`.

   :n:`with @bindings`  (in :n:`@one_term_with_bindings`)
      Provides explicit instances for
      the :term:`dependent premises <dependent premise>` of the type of
      :token:`one_term`.

   :n:`@occurrences`
     Selects specific subterms of the goal and/or hypotheses to apply
     the tactic to.  See :ref:`Occurrence clauses <occurrenceclauses>`.
     If it occurs in the :n:`@induction_principle`, then
     there can only be one :n:`@induction_clause`, which can't have its
     own :n:`@occurrences` clause.

   :n:`@induction_principle`
     Makes the tactic equivalent to
     :tacn:`induction` :n:`{+, @induction_clause } @induction_principle`.

   .. tacn:: edestruct {+, @induction_clause } {? @induction_principle }

      If the type of :n:`@one_term` (in :n:`@induction_arg`) has
      :term:`dependent premises <dependent premise>` or dependent premises
      whose values are not inferrable from the :n:`with @bindings` clause,

>       :n:`edestruct` turns them into existential variables to be resolved later on.

Reviewer: This paragraph is sufficient alone. You can remove the next one.

Author: Whoops, right.

Applying constructors
---------------------

The tactics presented here specialize :tacn:`apply` and
:tacn:`eapply` to constructors of inductive types.

.. tacn:: constructor {? @nat_or_var } {? with @bindings }

   First does :n:`repeat intro; hnf` on the goal.  If the result is an inductive
   type, then apply the appropriate constructor(s), and otherwise fail.
   If :n:`@nat_or_var` is specified and has the
   value `i`, it uses :n:`apply c__i`, where :n:`c__i` is the i-th constructor
   of :g:`I`.  If not specified, the tactic tries all the constructors,
   which can result in more than one success (e.g. for `\\/`) when using compound
   backtracking tactics such as `constructor; ...`.

   :n:`{? with @bindings }`
     If specified, the :n:`apply` is done as :n:`apply … with @bindings`.

     .. warning::

        The terms in :token:`bindings` are checked in the context
        where constructor is executed and not in the context where :tacn:`apply`
        is executed (the introductions are not taken into account).

   .. exn:: Not an inductive product.
      :undocumented:

   .. exn:: Not enough constructors.
      :undocumented:

>    .. exn: The type has no constructors.

Reviewer: ```suggestion
   .. exn:: The type has no constructors.
```

      :undocumented:

.. tacn:: split {? with @bindings }

   Equivalent to :n:`constructor 1 {? with @bindings }` when the
   conclusion is an inductive type with a single constructor.  The :n:`@bindings`
   specify any parameters required for the constructor. It is
   typically used to split conjunctions in the conclusion such as `A /\\ B` into
   two new goals `A` and `B`.

.. tacn:: exists {*, @bindings }

>    Equivalent to :n:`intros` followed by :n:`constructor 1 {? with binding__i }`

Reviewer: Actually `constructor` already includes the `intros` and `exists` behave the same (the bindings are understood before the `intros`). So to change to:
```suggestion
   Equivalent to :n:`constructor 1 {? with @bindings__i }`
```

Author: Equivalent to :n:`constructor 1 with @bindings__i` for each binding
   (or just :n:`constructor 1` if there are no :n:`@bindings`)

Reviewer: The fact that this error message can also be triggered by `split`. But not very important.

>    for each binding when the conclusion is an

Reviewer: Not for each binding, but for each set of `@bindings`.

Dummy example:

```coq
Inductive foo P : Prop :=
  bar : forall (n : nat) (m : nat), P n m -> foo P.

Goal foo (fun x y => foo (fun x' y' => x + x' = y + y')).
exists 0 0, 1 1. (* equivalent to constructor 1 with 0 0; constructor 1 with 1 1. *)
```

   inductive type with a single constructor.  It is typically used on
   existential quantifications in the form `exists x, P x.`

.. exn:: Not an inductive goal with 1 constructor.

>    :undocumented:

Reviewer: I don't like having error messages like this non-indented because it's not clear to what tactic they are attached. I would prefer to see it attached to one of the two tactics and the other one refer to the error message with `:exn:`.

Author: I indented it.  Not sure what you have in mind for "... the other one refer to the error message with `:exn:`".

Reviewer: The fact that this error message can also be triggered by `split`. Therefore, the description of `split` could contain a reference to this error message. But not very important.

.. tacn:: left {? with @bindings }
          right {? with @bindings }

   These tactics apply only if :g:`I` has two constructors, for
   instance in the case of a disjunction `A \\/ B`.
   Then they are respectively equivalent to
   :n:`constructor 1 {? with @bindings }` and
   :n:`constructor 2 {? with @bindings }`.

   .. exn:: Not an inductive goal with 2 constructors.
      :undocumented:

   .. tacn:: econstructor {? @nat_or_var {? with @bindings } }
             eexists {*, @bindings }
             esplit {? with @bindings }
             eleft {? with @bindings }

>              eright {? with @bindings }

Reviewer: Remove the indentation here, otherwise it looks like these are variants of `left` and `right`.

>       These tactics and their variants behave like :tacn:`constructor`,

Reviewer: ```suggestion
      These tactics behave like :tacn:`constructor`,
```

   .. todo PR Need more explanation about the tactic does when destruction_arg is
      a one_term that's not just an identifier'

   :n:`{+, @induction_clause }`
     Giving multiple :n:`@induction_clause`\s is equivalent to applying :n:`destruct`
     serially on each :n:`@induction_clause`.

   :n:`@natural`
     Equivalent to :tacn:`intros` :n:`until @natural` followed by `destruct` applied to
     the last introduced hypothesis.

     .. note::
        To destruct a number, use the syntax :n:`destruct (@natural)` (not
        very interesting anyway).

   :n:`@ident` (in :n:`@destruction_arg`)
     If :token:`ident` is a quantified variable in the conclusion
     of the goal, then :n:`destruct @ident` behaves like :tacn:`intros`
     :n:`until @ident; destruct @ident`.
     If :token:`ident` is a hypothesis of the context and :token:`ident`
     is no longer dependent in the goal after application
     of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as
     in :n:`destruct (@ident)`).

   :n:`as @induction_ipat`
      Provides names for the variables introduced
      into the context.  The :token:`or_and_intropattern` must have the

>       form :n:`[ {*| {* @intropattern } } ]` with one :n:`{* @intropattern }`

Reviewer: With this formulation, it's not so clear that only this form is accepted (and not an `equality_intropattern` or an `ident`).

Author: It appears that `induction_clause` (`as_or_and_ipat` in the mlgs) never accepts `equality_intropattern` (Looking at (e)case, (e)destruct, (e)induction and * inversion*).

IIUC, the change in cf4105502388e437c1cf361b5c3ddd8a482eef04 (allowing `equality_intropattern` in `as_or_and_ipat`) was not needed; the added production is never semantically permitted.  @herbelin?

Also, an `identref` is not accepted for these tactics (in `or_and_intropattern_loc`).  Since the `identref` is only reachable from these tactics, that production can be dropped from the grammar, in which case `or_and_intropattern_loc` can also be dropped, substituting `or_and_intropattern` in `as_or_and_ipat`.

The deleted productions (before dropping `or_and_intropattern_loc`:

or_and_intropattern_loc: [
| or_and_intropattern
~| identref~
]

as_or_and_ipat: [
| "as" or_and_intropattern_loc
~| "as" equality_intropattern~
|
]

I can put in a PR to do this if that's agreeable.

Reviewer: > IIUC, the change in [cf41055](https://github.com/coq/coq/commit/cf4105502388e437c1cf361b5c3ddd8a482eef04) (allowing `equality_intropattern` in `as_or_and_ipat`) was not needed; the added production is never semantically permitted.

I get it now! Thanks for digging this commit out. This was as an answer to Ralf in #13205, who was surprised that `destruct` didn't accept the injection intropattern. Therefore, this really is an implementation detail to improve user messages but should not appear in the doc.

Author: I simplified the grammar (dropping `induction_ipat`) and reworded the description:

   :n:`as @or_and_intropattern`
      Provides names for the variables introduced in each new subgoal.  The
      :token:`or_and_intropattern` must have one :n:`{* @intropattern }`
      for each constructor, given in the order in which the constructors are
      defined.  If there are not enough names, Coq picks fresh names.

      for each constructor in the order in which the constructors are defined.
      The items in
      each :n:`{* @intropattern }` are used to name the variables created in
      each new goal (one per constructor).  If there are not enough names,
      Coq picks fresh names.

   :n:`eqn : @naming_intropattern`
      Generates a new hypothesis in each new subgoal that is an equality between
      the term being case-analyzed and the associated constructor (applied to
      its arguments).  The name of the new item may be specified in the
      :n:`@naming_intropattern`.

   :n:`with @bindings`  (in :n:`@one_term_with_bindings`)
      Provides explicit instances for
      the :term:`dependent premises <dependent premise>` of the type of
      :token:`one_term`.

   :n:`@occurrences`
     Selects specific subterms of the goal and/or hypotheses to apply
     the tactic to.  See :ref:`Occurrence clauses <occurrenceclauses>`.
     If it occurs in the :n:`@induction_principle`, then

>      there can only be one :n:`induction_clause`, which can't have its
>      own :n:`occurrences` clause.

Reviewer: ```suggestion
     there can only be one :n:`@induction_clause`, which can't have its
     own :n:`@occurrences` clause.
```

>    :n:`using @one_term {? with @bindings }`
>      Makes the tactic equivalent to :tacn:`induction` :n:`{+, @induction_clause }`.

Reviewer: ```suggestion
   :n:`@induction_principle`
     Makes the tactic equivalent to :tacn:`induction` :n:`{+, @induction_clause } @induction_principle`.
```

   .. tacn:: edestruct {+, @induction_clause } {? @induction_principle }

      Equivalent to :n:`destruct @term` except that it does not
      fail if the instance of a dependent premises of the type
      of :token:`term` is not inferrable. Instead, the unresolved instances
      are left as existential variables to be inferred later, in the same way
      they are in :tacn:`eapply`.

.. tacn:: case {+, @induction_clause } {? @induction_principle }

   .. todo PR: clear as mud? (below) the syntax accepts more parameters than elim

>       not sure what to say here

Reviewer: Are all the left over `todo PR` still current?

Author: At line 206, for `case`, the description is pretty thin.  Maybe we can make it more descriptive?

Separately, shall I add `dependent` to the glossary as a synonym of `dependent product`?  (Note I used :term:`dependent <dependent product>` under `induction`, line 247.

The others are after line 250, so don't need to be addressed yet.

Reviewer: >  Maybe we can make it more descriptive?

I don't know enough about this tactic to suggest what to add.

> Separately, shall I add `dependent` to the glossary as a synonym of `dependent product`? (Note I used :term:`dependent <dependent product>` under `induction`, line 247.

I didn't review that part yet (because I had misunderstood where to stop) but I don't find this sentence very clear and I'm not comfortable making these synonyms.

   `case` only modifies the goal; it does not modify the :term:`local context`.

   .. tacn:: ecase {+, @induction_clause } {? @induction_principle }

      If the type of :n:`@one_term` (in :n:`@destruction_arg`) has
      :term:`dependent premises <dependent premise>`, or dependent premises
      whose values are not inferrable from the :n:`with @bindings` clause,
      :n:`ecase` turns them into existential variables to be resolved later on.

   .. tacn:: case_eq @one_term

      A variant of the :n:`case` tactic that allows
      performing case analysis on a term without completely forgetting its original
      form. This is done by generating equalities between the original form of the
      term and the outcomes of the case analysis.  We recommend using the
      :tacn:`destruct` tactic with an `eqn:` clause instead.

.. tacn:: casetype @one_term
   :undocumented:

.. tacn:: simple destruct {| @ident | @natural }

   Equivalent to :tacn:`intros` :n:`until {| @ident | @natural }; case @ident`
   where :n:`@ident` is a quantified variable of the goal and otherwise fails.

.. tacn:: dependent destruction @ident {? generalizing {+ @ident } } {? using @one_term }

>    :undocumented:

Reviewer: Maybe add a reference to the example section at the bottom...

   If :n:`@nat_or_var` is specified and has the
   value `i`, it uses :n:`apply c__i`, where :n:`c__i` is the i-th constructor
   of :g:`I`.  If not specified, the tactic tries all the constructors,
   which can result in more than one match (e.g. for `\\/`) in compound backtracking
   tactics such as `constructor; ...` is used.

   :n:`{? with @bindings }`
     If specified, the :n:`apply` is done as :n:`apply … with @bindings`.

     .. warning::

        The terms in :token:`bindings` are checked in the context
        where constructor is executed and not in the context where :tacn:`apply`
        is executed (the introductions are not taken into account).

   .. exn:: Not an inductive product.
      :undocumented:

   .. exn:: Not enough constructors.
      :undocumented:

   .. todo PR: maybe split, left, right etc should be at the left margin?

   .. tacn:: split {? with @bindings }

      Equivalent to :n:`constructor 1 {? with @bindings }` when the
      conclusion is an inductive type with a single constructor. It is
      typically used to split conjunctions in the conclusion such as `A /\\ B` into
      two new goals `A` and `B`.

>       .. tacn:: exists {*, @bindings }

Author: "a shared description" - do you mean to put these lines together and meld the text? Or something else?

```
.. tacn:: split {? with @bindings }
.. tacn:: exists {*, @bindings }
```

>       .. tacn:: exists {*, @bindings }

Author: I'm not seeing how to meld them in a helpful way.  I tweaked them slightly; `exists` was a bit incorrect.  I think they are correct now.  Can you elaborate on how you would meld them?

FWIW, `split` does a bit more than advertised:

```
Goal forall A B, A /\ B -> True.
split.   (* No more subgoals. *)
```

>       .. tacn:: exists {*, @bindings }

Reviewer 2: > FWIW, split does a bit more than advertised:

This is because `True` has only one constructor with no arguments, hence `constructor 1` performs an implicit `intros` and then solves the goal.  Is this missing from the advertising?

             eexists {*, @bindings }
             esplit {? with @bindings }
             eleft {? with @bindings }
             eright {? with @bindings }

      These tactics and their variants behave like :tacn:`constructor`,
      :tacn:`exists`, :tacn:`split`, :tacn:`left` and :tacn:`right`,
      but they introduce existential variables instead of failing
      when a variable can't be instantiated
      (cf. :tacn:`eapply` and :tacn:`apply`).

.. example:: :tacn:`constructor`, :tacn:`left` and :tacn:`right`

   .. coqtop:: reset all

      Print or.  (* or, represented by \/, has two constructors, or_introl and or_intror *)
      Goal  forall P1 P2 : Prop, P1 -> P1 \/ P2.
      constructor 1.  (* equivalent to "left" *)
      apply H.  (* success *)

   In contrast, we won't be able to complete the proof if we select constructor 2:

   .. coqtop:: reset none

      Goal  forall P1 P2 : Prop, P1 -> P1 \/ P2.

   .. coqtop:: all

      constructor 2.  (* equivalent to "right" *)

   You can also apply a constructor by name.  (`apply or_intror` will fail):

   .. coqtop:: reset none

      Goal  forall P1 P2 : Prop, P1 -> P1 \/ P2.

   .. coqtop:: all

>       apply or_introl.  (* equivalent to "left" *)

Reviewer: ```suggestion
      intros; apply or_introl.  (* equivalent to "left" *)
```

If you do not do the `intros`, then it is not equivalent to `constructor 1` / `left`.

With the `intros`, it doesn't solve the goal and `apply or_intror` doesn't fail (it just leads to a dead-end).

.. _CaseAnalysisAndInduction:

Case analysis
-------------

The tactics in this section implement case

> analysis on inductive or co-inductive objects (see :ref:`gallina-inductive-definitions`).

Reviewer: In fact, the most relevant section to refer to at this point is "Variants and the `match` construct" (since this is the section where recursion doesn't matter).

Some useful terms:

* `n` is a :gdef:`dependent premise` of `forall n:nat, n + 0 = n` because
  `n` appears both in the binder of the `forall` and in the quantified statement
  `n + 0 = n`.

* `A` and `B` are :gdef:`non-dependent premises <non-dependent premise>` of `A -> B -> C`
  because they don't appear in a `forall` binder.
  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard

>   `_` can't be referred to in the quantified statement.)

Reviewer: These are terms of general interest. Therefore, it would make sense to move them to some earlier section of the refman. I'm thinking here: https://coq.github.io/doc/master/refman/language/core/assumptions.html#functions-fun-and-function-types-forall

Author: Will do.  I'd like to add "dependent product" and "product", too, if you can provide definitions.

Reviewer: I would suggest using the term "product type" instead of just "product" for clarity. Product types are just `forall @ident : @type__1, @type__2` by definition (dependent meaning, again, that `@ident` appears in `@type__2`). They can also be called "function types" given that they correspond exactly to the types of functions (or abstractions as this paragraph call them). Indeed, the type of `fun @ident : @type__1 ==> @term` is `forall @ident : @type__1, @type__2` where `@type__2` is the type of `@term`.

Author: Are "abstraction" and "generalization" the same thing?  If so, I think I'd prefer the latter here because we do have a `generalize` tactic.  But both terms are quite common and rarely defined.

Reviewer 2: To my ear, abstraction has a slight connotation of being over a variable, and generalization has the connotation of being over a term.  (That is, abstraction emphasizes the part where you introduce a lambda/pi to close over the abstracted variable, whereas generalization emphasizes the part where you replace a term (which might or might not be just a variable) with a fresh variable.). Or perhaps in some contexts "abstraction" suggests introducing a lambda while "generalization" suggests introducing a forall. But, yes, they are basically the same, and these differences in connotations are very slight.  And I would like Coq to get a reduction strategy called "abstract over", which performs `pattern` without creating the application(s) at the end.

Author: I left abstraction alone for now.  `generalize` is not quite what I remembered.

.. tacn:: destruct @induction_clause_list

   .. insertprodn induction_clause_list as_or_and_ipat

   .. prodn::

>       induction_clause_list ::= {+, @induction_clause } {? using @one_term {? with @bindings } } {? @occurrences }

Reviewer: To make things more readable, I suggest:

1. Move the trailing `{? @occurrences }` inside the optional `using` clause. When the `using` clause` is not provided, there is anyway some ambiguity wrt which parsing path is selected in the current presentation.
2. Introduce a new nonterminal for `using @one_term {? with @bindings } {? @occurrences }`.
3. Now that the definition of `induction_clause_list` is much shorter (reduced to `{+, @induction_clause } {? induction_principle }`), it becomes reasonable to splice it.

Author: What's a better name for `destruction_arg`?

Author: ![image](https://user-images.githubusercontent.com/1253341/122153210-5de83200-ce17-11eb-8c2e-668bd87c9875.png)

Reviewer: I don't have great ideas. This could become `term_or_position_argument`. Or maybe it would make sense to splice it and simply replace it everywhere with `{| @natural | @one_term_with_bindings }`...

Author: `induction_arg` would be a little better, at least similar to the other NT names seen here.

      induction_clause ::= @destruction_arg {? @as_or_and_ipat } {? eqn : @naming_intropattern } {? @occurrences }
      destruction_arg ::= @natural
      | @one_term_with_bindings
      as_or_and_ipat ::= as {| @or_and_intropattern | @ident }
      | as @equality_intropattern

>    Generates a subgoal for each constructor of the inductive or co-inductive

Reviewer: Should probably say "Perform case analysis" before "generates a subgoal for each constructor".

   type in the :n:`@one_term` in :n:`@destruction_arg`.  Unlike :tacn:`induction`,
   no induction hypothesis is generated by :tacn:`destruct`.

   The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
   the tactic checks that all subterms matching the pattern in the conclusion
   and the hypotheses are compatible and performs case analysis using this subterm.

   .. todo PR Need more explanation about the tactic does when destruction_arg is

>       a one_term that's not just an identifier'

Reviewer: Example from a recent proof I did:

```coq
destruct (compare a1 b1) eqn:Hab
```

Does a case analysis on `compare a1 b1` (can be `Eq`, `Lt` or `Gt`). For each case, thanks to the `eqn:` clause, it will generate a new hypothesis of the form `Hab : compare a1 b1 = Eq (* or Lt or Gt *)`. In this case, the `one_term` doesn't even need to appear in the goal or the hypotheses.

Author: Interesting.  I didn't realize that `eqn:` would create additional hypotheses.

Without eqn, the behavior is to create similar equations for an inductive-typed value and then substitute the rhs wherever the lhs appeared in the context?

Separately, what exactly is ` { ... } + { ... }`?  It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.

1 subgoal
n : nat
IHn : forall m0 : nat, {n = m0} + {n <> m0}
m : nat
______________________________________(1/1)
{S n = S m} + {S n <> S m}

1: {
 
![image](https://user-images.githubusercontent.com/1253341/122143083-8f580200-ce05-11eb-9ead-eeb289455b84.png)

2 : {
 
![image](https://user-images.githubusercontent.com/1253341/122143092-954de300-ce05-11eb-867d-65d3d02adf04.png)

Reviewer 2: > Separately, what exactly is { ... } + { ... }?

A notation for Coq.Init.Datatypes.sumbool, I believe.  (Might be in Logic or Specif rather than Datatype.). Note that you can answer most of these questions yourself by doing Locate or About on the notation string and then Locate again on the identifier (so `Locate "{ _ } + { _ }". Locate sumbool.` in this case)

Reviewer: Or just `About "{ _ } + { _ }".`

>Without eqn, the behavior is to create similar equations for an inductive-typed value and then substitute the rhs wherever the lhs appeared in the context?

Yes. Note that with `eqn`, the substitution also occurs. But it's still useful to get the equation for later.

>  It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.

Yes, I don't know why it is so.

Reviewer 2: > > It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.
> 
> Yes, I don't know why it is so.

`e` from the first letter of the type `eq`, `n0` from the first letter of the type `not`, I think

Author: The curly braces mean that it's a set?

Further off track: FWIW, I don't understand what `Specify.v` is useful for.  Some of the comments make me think it's partly about manipulating dependent types--something I tried to figure out a long while ago but didn't get anywhere.  (Not the concepts, but the nuts and bolts of using them.)

For example:

```
(** Projections of [sig]

    An element [y] of a subset [{x:A | (P x)}] is the pair of an [a]
    of type [A] and of a proof [h] that [a] satisfies [P].  Then
    [(proj1_sig y)] is the witness [a] and [(proj2_sig y)] is the
    proof of [(P a)] *)
```

Reviewer 2: > The curly braces mean that it's a set?

Curly braces seem to indicate turning a `Prop` into a `Set`
https://github.com/coq/coq/blob/147f4da1beee028be550f098401398f5822af78e/theories/Init/Specif.v#L887-L890

> Further off track: FWIW, I don't understand what `Specify.v` is useful for.

I believe `Specif.v` is short for "specifications" (which indeed are largely based on using dependent types)

Author: Thanks.  I expect the standard library is not well documented in general.  Something to look at later.

   :n:`{+, @induction_clause }`
     Giving multiple :n:`@induction_clause`\s is equivalent to applying :n:`destruct`

>      serially on each :n:`@induction_clause`.

Reviewer: I think "chaining calls to `destruct` for each `@induction_clause`" would read better.

Author: I don't quite see it.  "chaining" is not precisely defined.  The dictionary doesn't have a verb definition similar to this (not that that's definitive).

Reviewer: OK maybe it's a Frenchism. We often say "chaîner" for `;`.

Reviewer 2: Not a Frenchism.  In English it comes from a chain as in 
![image](https://user-images.githubusercontent.com/396076/122256918-7d29a280-ce9d-11eb-8350-201d0bfd6d6f.png)

"chaining" is linking or sequencing multiple things, as if in a chain.

Author: My comment was not that "chaining" isn't used somewhat informally in this sense, it was that "chaining" is not precise.

In basic calculus: "The chain rule states that the derivative of f(g(x)) is f'(g(x))⋅g'(x)."

https://web.mit.edu/wwmath/calculus/differentiation/chain.html: "This rule is called the chain rule because we use it to take derivatives of composites of functions by chaining together their derivatives."  Clearly quite different from `destruct`!

"chaîner" appears to be closer to "linked", I think.

```
chaîner
verb
 to be chained to another data item
```

Reviewer 2: >  Clearly quite different from destruct!

"chaining" is not about `destruct` ; IIUC, it's about the `;` in the fact that `destruct a, b` is the same as `destruct a; destruct b`.

Reviewer 2: Also, `;` is composition of tactics in a way that's very similar to composition of functions, and this similarity can be made technically precise.  In this way the "chain" in the chain rule is about the same thing as the "chain" in "chaining tactic calls"

Author: > >  Clearly quite different from destruct!

> "chaining" is not about `destruct` ; IIUC, it's about the `;` in the fact that `destruct a, b` is the same as `destruct a; destruct b`.

I was referring to `destruct a, b` of course, I just didn't spell it out completely.

> this similarity can be made technically precise.  In this way the "chain" in the chain rule is about the same thing as the "chain" in "chaining tactic calls"

Maybe so, but it doesn't make sense to define that term here--it's simpler just to explain `destruct a,b` with a short phrase.  "Chaining" is used in 9 other places in the doc.  If we make it a defined term, the other uses need to examined to make sure the definition fits them.

Not obvious to me how you would define it in a useful way.

Fun as it is to bikeshed small details, I would like to move on.  @Zimmi48 if you want to say "chained" here then I'll do that, but I do think that's vague.

Reviewer: I'm quite fine with the current state, even though I find the terminology "compound tactic" less clear than "chaining tactics" (none of them being clearly defined in the manual yet).

Author: I dropped "compound tactic" and added hyperlinks.

     If :token:`ident` is a quantified variable in the conclusion
     of the goal, then :n:`destruct @ident` behaves like :tacn:`intros`
     :n:`until @ident; destruct @ident`.
     If :token:`ident` is a hypothesis of the context and :token:`ident`
     is no longer dependent in the goal after application
     of :tacn:`destruct`, it is erased (to avoid erasure, use parentheses, as
     in :n:`destruct (@ident)`).

   :n:`using @one_term {? with @bindings }`
     Makes the tactic equivalent to :tacn:`induction` :n:`@induction_clause_list`.

   :n:`@as_or_and_ipat`
      Provides names for the variables introduced
      into the context.  The :token:`as_or_and_ipat` must have the
      form :n:`[ {*| {* @intropattern } } ]` with one :n:`{* @intropattern }`
      for each constructor in the order in which the constructors are defined.
      The items in
      each :n:`{* @intropattern }` are used to name the variables created in
      each new goal (one per constructor).  If there are not enough names,
      Coq picks fresh names.

   :n:`eqn : @naming_intropattern`
      Defines new items in the context for the matching terms using the specified

>       or generated names.

Reviewer: This is very unclear. What it should say is that it generates a new hypothesis which will be an equality between the term being case-analyzed and the constructor (applied to its arguments) corresponding to the current case. See my example above.

Author: :n:`eqn : @naming_intropattern`
      Generates a new hypothesis in each new subgoal that is an equality between
      the term being case-analyzed and the associated constructor (applied to
      its arguments).  The name of the new item may be specified in the
      :n:`@naming_intropattern`.

   :n:`with @bindings`  (in :n:`@one_term_with_bindings`)
      Provides explicit instances for
      the :term:`dependent premises <dependent premise>` of the type of
      :token:`one_term`.

   :n:`@occurrences`
     Selects specific subterms of the goal and/or hypotheses to apply
     the tactic to.  See :ref:`Occurrence clauses <occurrenceclauses>`.
     If it occurs at the end of the :n:`@induction_clause_list`, then
     there can only be one :n:`induction_clause`, which can't have its

>      own :n:`occurrences` clause.

Reviewer: This bit goes away by applying my suggestion above of moving this trailing occurrence clause inside the `using` clause.

Author: I don't think so.  IIUC each `induction_clause` can have it's own `occurrences`, for example
`destruct n at 1, n at 1.` is valid.

I created #14510

   .. tacn:: edestruct @induction_clause_list

      Equivalent to :n:`destruct @term` except that it does not
      fail if the instance of a dependent premises of the type
      of :token:`term` is not inferrable. Instead, the unresolved instances
      are left as existential variables to be inferred later, in the same way
      they are in :tacn:`eapply`.

.. tacn:: case @induction_clause_list

   .. todo PR: clear as mud? (below) the syntax accepts more parameters than elim
      not sure what to say here

   An older, more basic tactic to perform case analysis without
   recursion. It behaves like :tacn:`elim` but uses a case-analysis

>    elimination principle rather than a recursive one.

Reviewer: I suggest dropping the sentence with the reference to `elim` completely. However, I would add one to say that we recommend `destruct` when possible.

   `case` only modifies the goal; it does not modify the :term:`local context`.

> .. tacn:: ecase @induction_clause_list

Reviewer: Can you indent this to make it appear as a variant?

   If the type of :n:`@one_term` (in :n:`@destruction_arg`) has
   :term:`dependent premises <dependent premise>`, or dependent premises
   whose values are not inferrable from the :n:`with @bindings` clause,
   :n:`ecase` turns them into existential variables to be resolved later on.

.. tacn:: casetype @one_term
   :undocumented:

.. tacn:: simple destruct {| @ident | @natural }

   Equivalent to :tacn:`intros` :n:`until {| @ident | @natural }; case @ident`
   where :n:`@ident` is a quantified variable of the goal and otherwise fails.

.. tacn:: dependent destruction @ident {? generalizing {+ @ident } } {? using @one_term }
   :undocumented:

> .. tacn:: case_eq @one_term

Reviewer: You can add to the description of this tactic that we recommend using `destruct` with the `eqn` clause instead. Maybe you can also move it to make it appear as a variant of `case`.

.. seealso:: :tacn:`functional induction`

.. _equality-inductive_types:

:tacn:`discriminate` and :tacn:`injection`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: If you don't mind, let's also keep this new subsection title for the next PR (because for now it doesn't correctly reflect the content below---given that `inversion`, `fix` and `cofix` are also there). Since you need the label, I suggest moving it down above the "Equality and inductive sets" section title.
